File: Boutique-backend/src/main/java/com/sapthala/service/WorkflowService.java
package com.sapthala.service;

import com.sapthala.model.*;
import com.sapthala.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class WorkflowService {
    
    private final TaskRepository taskRepository;
    private final StaffRepository staffRepository;
    private final WorkflowTemplateRepository workflowTemplateRepository;
    private final NotificationRepository notificationRepository;
    
    /**
     * Creates workflow tasks for a new order
     */
    public void createWorkflowForOrder(String orderId, String category, String subcategory, 
                                      String garmentName, String customerName, String customerPhone,
                                      String deliveryDate) {
        try {
            // Find appropriate workflow template
            WorkflowTemplate template = findWorkflowTemplate(category, subcategory);
            if (template == null) {
                log.error("No workflow template found for category: {} subcategory: {}", category, subcategory);
                return;
            }
            
            log.info("Creating workflow for Order: {}", orderId);
            
            // Create tasks for each stage
            List<Task> tasks = new ArrayList<>();
            for (WorkflowTemplate.WorkflowStage stage : template.getStages()) {
                Task task = Task.builder()
                    .taskId("TASK-" + System.currentTimeMillis() + "-" + stage.getIndex())
                    .orderId(orderId)
                    .orderDetails(Task.OrderDetails.builder()
                        .customerName(customerName)
                        .customerPhone(customerPhone)
                        .garmentName(garmentName)
                        .deliveryDate(deliveryDate)
                        .build())
                    .stage(stage.getName())
                    .stageIndex(stage.getIndex())
                    .description(stage.getDescription())
                    .status("pending")
                    .estimatedDuration(stage.getEstimatedTime())
                    .notes("")
                    .priority("normal")
                    .statusHistory(Arrays.asList(Task.StatusHistory.builder()
                        .status("pending")
                        .timestamp(LocalDateTime.now())
                        .changedBy("SYSTEM")
                        .build()))
                    .photos(new ArrayList<>())
                    .createdAt(LocalDateTime.now())
                    .updatedAt(LocalDateTime.now())
                    .build();
                
                tasks.add(task);
            }
            
            taskRepository.saveAll(tasks);
            log.info("✅ Created {} tasks for Order: {}", tasks.size(), orderId);
            
            // Auto-assign first task
            if (!tasks.isEmpty()) {
                assignTaskToAvailableStaff(tasks.get(0));
            }
            
        } catch (Exception e) {
            log.error("Error creating workflow for order: {}", orderId, e);
        }
    }
    
    /**
     * Smart auto-assignment logic
     */
    public void assignTaskToAvailableStaff(Task task) {
        try {
            // Get required role from task
            String requiredRole = getRequiredRole(task.getStage());
            
            // Find available staff with required role
            Staff assignedStaff = findBestStaffMember(requiredRole);
            
            if (assignedStaff != null) {
                task.setAssignedTo(Task.StaffAssignment.builder()
                    .staffId(assignedStaff.getStaffId())
                    .name(assignedStaff.getName())
                    .phone(assignedStaff.getPhone())
                    .role(assignedStaff.getRole())
                    .build());
                task.setStatus("assigned");
                
                taskRepository.save(task);
                
                // Update staff workload
                updateStaffWorkload(assignedStaff.getStaffId(), 1);
                
                // Send notification
                sendTaskNotification(assignedStaff.getStaffId(), task);
                
                log.info("✅ Task {} assigned to staff: {}", task.getTaskId(), assignedStaff.getName());
            } else {
                log.warn("⚠️ No available staff found for role: {}", requiredRole);
            }
        } catch (Exception e) {
            log.error("Error assigning task: {}", task.getTaskId(), e);
        }
    }
    
    /**
     * Find best available staff member based on:
     * - Availability status
     * - Current workload
     * - Quality rating
     * - Experience
     */
    private Staff findBestStaffMember(String requiredRole) {
        List<Staff> availableStaff = staffRepository
            .findByRoleAndIsActiveTrueAndAvailabilityStatus(requiredRole, "available");
        
        if (availableStaff.isEmpty()) {
            return null;
        }
        
        // Sort by: workload (ascending), quality rating (descending), experience (descending)
        return availableStaff.stream()
            .sorted(Comparator
                .comparingInt((Staff s) -> s.getCurrentWorkload().getActiveTasks()) // Less workload first
                .thenComparingDouble((Staff s) -> -s.getStatistics().getQualityRating()) // Higher rating first
                .thenComparingInt((Staff s) -> -s.getStatistics().getTotalTasksCompleted())) // More experience first
            .findFirst()
            .orElse(null);
    }
    
    /**
     * Update task status (Start, Pause, Resume, End)
     */
    public void updateTaskStatus(String taskId, String newStatus, String staffId) {
        try {
            Optional<Task> optionalTask = taskRepository.findByTaskId(taskId);
            if (optionalTask.isEmpty()) {
                log.error("Task not found: {}", taskId);
                return;
            }
            
            Task task = optionalTask.get();
            String previousStatus = task.getStatus();
            
            // Update task status
            task.setStatus(newStatus);
            task.setUpdatedAt(LocalDateTime.now());
            
            // Record status change
            if (task.getStatusHistory() == null) {
                task.setStatusHistory(new ArrayList<>());
            }
            task.getStatusHistory().add(Task.StatusHistory.builder()
                .status(newStatus)
                .timestamp(LocalDateTime.now())
                .changedBy(staffId)
                .build());
            
            // Handle status-specific logic
            switch (newStatus) {
                case "in-progress":
                    task.setStartedAt(LocalDateTime.now());
                    task.setPausedAt(null);
                    break;
                case "paused":
                    task.setPausedAt(LocalDateTime.now());
                    break;
                case "completed":
                    task.setCompletedAt(LocalDateTime.now());
                    // Calculate actual duration
                    if (task.getStartedAt() != null) {
                        long minutes = java.time.temporal.ChronoUnit.MINUTES
                            .between(task.getStartedAt(), task.getCompletedAt());
                        task.setActualDuration((int) minutes);
                    }
                    // Assign next task
                    assignNextTask(task);
                    break;
            }
            
            taskRepository.save(task);
            log.info("✅ Task {} status updated: {} → {}", taskId, previousStatus, newStatus);
            
        } catch (Exception e) {
            log.error("Error updating task status: {}", taskId, e);
        }
    }
    
    /**
     * Auto-assign next task when current task completes
     */
    private void assignNextTask(Task completedTask) {
        try {
            // Find all tasks for this order
            List<Task> orderTasks = taskRepository.findByOrderId(completedTask.getOrderId())
                .stream()
                .sorted(Comparator.comparingInt(Task::getStageIndex))
                .collect(Collectors.toList());
            
            // Find next pending task
            Task nextTask = orderTasks.stream()
                .filter(t -> t.getStatus().equals("pending") && t.getStageIndex() > completedTask.getStageIndex())
                .findFirst()
                .orElse(null);
            
            if (nextTask != null) {
                assignTaskToAvailableStaff(nextTask);
                log.info("✅ Next task auto-assigned: {}", nextTask.getTaskId());
            } else {
                log.info("✅ All tasks completed for order: {}", completedTask.getOrderId());
            }
        } catch (Exception e) {
            log.error("Error assigning next task", e);
        }
    }
    
    /**
     * Send notification to staff member
     */
    private void sendTaskNotification(String staffId, Task task) {
        try {
            Notification notification = Notification.builder()
                .notificationId("NOTIF-" + System.currentTimeMillis())
                .staffId(staffId)
                .type("new-task")
                .title("New Task Assigned")
                .body("Task: " + task.getDescription() + " for Order: " + task.getOrderId())
                .data(new HashMap<String, Object>() {{
                    put("taskId", task.getTaskId());
                    put("orderId", task.getOrderId());
                    put("stage", task.getStage());
                    put("priority", task.getPriority());
                    put("estimatedTime", task.getEstimatedDuration());
                }})
                .isRead(false)
                .createdAt(LocalDateTime.now())
                .expiresAt(LocalDateTime.now().plusDays(30))
                .build();
            
            notificationRepository.save(notification);
            log.info("✅ Notification sent to staff: {}", staffId);
        } catch (Exception e) {
            log.error("Error sending notification", e);
        }
    }
    
    /**
     * Update staff workload
     */
    private void updateStaffWorkload(String staffId, int change) {
        Optional<Staff> optionalStaff = staffRepository.findByStaffId(staffId);
        if (optionalStaff.isPresent()) {
            Staff staff = optionalStaff.get();
            int currentWorkload = staff.getCurrentWorkload().getActiveTasks();
            staff.getCurrentWorkload().setActiveTasks(currentWorkload + change);
            
            // Auto update availability status
            if (currentWorkload + change >= staff.getCurrentWorkload().getMaxConcurrentTasks()) {
                staff.getAvailability().setStatus("busy");
            } else if (currentWorkload + change == 0) {
                staff.getAvailability().setStatus("available");
            }
            staff.getAvailability().setLastUpdated(LocalDateTime.now());
            
            staffRepository.save(staff);
        }
    }
    
    /**
     * Get required role for a task stage
     */
    private String getRequiredRole(String stage) {
        return switch (stage) {
            case "dye" -> "dyer";
            case "cutting" -> "cutter";
            case "stitching" -> "tailor";
            case "qc" -> "qc";
            case "delivery" -> "delivery";
            default -> "general";
        };
    }
    
    /**
     * Find appropriate workflow template
     */
    private WorkflowTemplate findWorkflowTemplate(String category, String subcategory) {
        // Try to find specific template for category + subcategory
        Optional<WorkflowTemplate> template = workflowTemplateRepository
            .findByCategoryAndSubcategoryAndIsActiveTrue(category, subcategory);
        
        // Fallback to category-only template
        if (template.isEmpty()) {
            template = workflowTemplateRepository
                .findByCategoryAndIsActiveTrue(category);
        }
        
        // Fallback to universal template
        if (template.isEmpty()) {
            template = workflowTemplateRepository
                .findByCategoryAndIsActiveTrue("all");
        }
        
        return template.orElse(null);
    }
}

File: services/workflowService.js
const { Staff, Order, Notification, Settings } = require('../database');

class WorkflowService {
  
  // Auto-assign task to available staff
  static async autoAssignTask(order, stageId) {
    try {
      const settings = await Settings.findOne();
      const stage = settings.workflowStages.find(s => s.id === stageId);
      
      if (!stage) return null;

      // Find available staff with required skills
      const availableStaff = await Staff.find({
        workflowStages: stageId,
        isAvailable: true,
        currentTaskCount: { $lt: 3 } // Max 3 concurrent tasks
      }).sort({ currentTaskCount: 1, rating: -1 });

      if (availableStaff.length === 0) return null;

      const selectedStaff = availableStaff[0];
      
      // Find the task in the order
      const taskIndex = order.workflowTasks.findIndex(t => t.stageId === stageId);
      if (taskIndex === -1) return null;

      // Assign task
      order.workflowTasks[taskIndex].status = 'assigned';
      order.workflowTasks[taskIndex].assignedTo = selectedStaff._id;
      order.workflowTasks[taskIndex].assignedToName = selectedStaff.name;
      order.workflowTasks[taskIndex].updatedAt = new Date();

      // Update staff task count
      selectedStaff.currentTaskCount += 1;
      await selectedStaff.save();
      await order.save();

      // Send notification
      await this.sendTaskNotification(selectedStaff, order, stage, 'assigned');

      return selectedStaff;
    } catch (error) {
      console.error('Auto-assign task error:', error);
      return null;
    }
  }

  // Progress to next workflow stage
  static async progressToNextStage(order, completedStageId) {
    try {
      const settings = await Settings.findOne();
      const currentStage = settings.workflowStages.find(s => s.id === completedStageId);
      
      if (!currentStage) return;

      // Use only the stages actually selected for this order
      const orderStageIds = (order.workflowTasks || []).map(t => t.stageId);
      const orderStages = settings.workflowStages.filter(s => orderStageIds.includes(s.id))
          .sort((a, b) => a.order - b.order);
      const currentIdx = orderStages.findIndex(s => s.id === completedStageId);
      const nextStage = orderStages[currentIdx + 1] || null;
            
      if (!nextStage) {
        // All stages completed
        order.status = 'completed';
        order.currentStage = 'completed';
        await order.save();
        return;
      }

      // Activate next stage
      const nextTaskIndex = order.workflowTasks.findIndex(t => t.stageId === nextStage.id);
      if (nextTaskIndex !== -1) {
        order.workflowTasks[nextTaskIndex].status = 'pending';
        order.workflowTasks[nextTaskIndex].updatedAt = new Date();
        order.currentStage = nextStage.id;
        await order.save();

        // Auto-assign next task
        await this.autoAssignTask(order, nextStage.id);
      }
    } catch (error) {
      console.error('Progress to next stage error:', error);
    }
  }

  // Send task notification
  static async sendTaskNotification(staff, order, stage, type) {
    try {
      let title, message;
      
      switch (type) {
        case 'assigned':
          title = `New ${stage.name} Task Assigned`;
          message = `Order #${order.orderId} for ${order.customerName} has been assigned to you.`;
          break;
        case 'completed':
          title = `${stage.name} Task Completed`;
          message = `Order #${order.orderId} ${stage.name} stage has been completed.`;
          break;
        default:
          title = 'Task Update';
          message = `Order #${order.orderId} has been updated.`;
      }

      // Save notification to database
      const notification = new Notification({
        type: 'task_assigned',
        title,
        message,
        recipientId: staff._id,
        orderId: order.orderId,
        taskId: `${order.orderId}_${stage.id}`
      });

      await notification.save();

      // Here you would integrate with FCM or WebSocket for real-time notifications
      console.log(`📱 Notification sent to ${staff.name}: ${title}`);
      
      return notification;
    } catch (error) {
      console.error('Send notification error:', error);
    }
  }

  // Get staff notifications
  static async getStaffNotifications(staffId, limit = 10) {
    try {
      const staff = await Staff.findOne({ staffId });
      if (!staff) return [];

      const notifications = await Notification.find({ 
        recipientId: staff._id 
      })
      .sort({ sentAt: -1 })
      .limit(limit);

      return notifications;
    } catch (error) {
      console.error('Get notifications error:', error);
      return [];
    }
  }

  // Mark notification as read
  static async markNotificationRead(notificationId) {
    try {
      await Notification.findByIdAndUpdate(notificationId, {
        isRead: true,
        readAt: new Date()
      });
    } catch (error) {
      console.error('Mark notification read error:', error);
    }
  }

  // Get staff tasks
  static async getStaffTasks(staffId, status = null) {
    try {
      const staff = await Staff.findOne({ staffId });
      if (!staff) return [];

      let query = {
        'workflowTasks.assignedTo': staff._id
      };

      if (status) {
        query['workflowTasks.status'] = status;
      }

      const orders = await Order.find(query);
      
      const tasks = [];
      orders.forEach(order => {
        order.workflowTasks.forEach(task => {
          if (task.assignedTo && task.assignedTo.toString() === staff._id.toString()) {
            if (!status || task.status === status) {
              tasks.push({
                ...task.toObject(),
                orderId: order.orderId,
                customerName: order.customerName,
                customerPhone: order.customerPhone,
                garmentType: order.garmentType,
                deliveryDate: order.deliveryDate,
                totalAmount: order.totalAmount
              });
            }
          }
        });
      });

      return tasks.sort((a, b) => new Date(b.updatedAt) - new Date(a.updatedAt));
    } catch (error) {
      console.error('Get staff tasks error:', error);
      return [];
    }
  }

  // Get available tasks for staff
  static async getAvailableTasksForStaff(staffId) {
    try {
      const staff = await Staff.findOne({ staffId });
      if (!staff) return [];

      const orders = await Order.find({
        'workflowTasks.status': 'pending',
        'workflowTasks.stageId': { $in: staff.workflowStages }
      });

      const availableTasks = [];
      orders.forEach(order => {
        order.workflowTasks.forEach(task => {
          if (task.status === 'pending' && staff.workflowStages.includes(task.stageId)) {
            availableTasks.push({
              ...task.toObject(),
              orderId: order.orderId,
              customerName: order.customerName,
              customerPhone: order.customerPhone,
              garmentType: order.garmentType,
              deliveryDate: order.deliveryDate,
              totalAmount: order.totalAmount
            });
          }
        });
      });

      return availableTasks.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    } catch (error) {
      console.error('Get available tasks error:', error);
      return [];
    }
  }

  // Update task status
  static async updateTaskStatus(orderId, stageId, staffId, newStatus, data = {}) {
    try {
      const order = await Order.findOne({ orderId });
      const staff = await Staff.findOne({ staffId });
      
      if (!order || !staff) return false;

      const taskIndex = order.workflowTasks.findIndex(t => t.stageId === stageId);
      if (taskIndex === -1) return false;

      const task = order.workflowTasks[taskIndex];
      
      // Update task based on status
      task.status = newStatus;
      task.updatedAt = new Date();

      switch (newStatus) {
        case 'assigned':
          task.assignedTo = staff._id;
          task.assignedToName = staff.name;
          staff.currentTaskCount += 1;
          break;
        case 'started':
          task.startedAt = new Date();
          break;
        case 'paused':
          task.pausedAt = new Date();
          if (data.notes) task.notes = (task.notes || '') + `\nPaused: ${data.notes}`;
          break;
        case 'resumed':
          task.resumedAt = new Date();
          break;
        case 'completed':
          task.completedAt = new Date();
          if (data.notes) task.notes = (task.notes || '') + `\nCompleted: ${data.notes}`;
          if (data.qualityRating) task.qualityRating = data.qualityRating;
          if (data.images) task.images = [...(task.images || []), ...data.images];
          
          // Calculate time spent
          if (task.startedAt) {
            let timeSpent = new Date() - new Date(task.startedAt);
            if (task.pausedAt && task.resumedAt) {
              timeSpent -= new Date(task.resumedAt) - new Date(task.pausedAt);
            }
            task.timeSpent = Math.round(timeSpent / (1000 * 60)); // in minutes
          }
          
          staff.currentTaskCount = Math.max(0, staff.currentTaskCount - 1);
          
          // Progress to next stage
          const settings = await Settings.findOne();
          const stage = settings.workflowStages.find(s => s.id === stageId);
          await this.progressToNextStage(order, stageId);
          await this.sendTaskNotification(staff, order, stage, 'completed');
          break;
      }

      await staff.save();
      await order.save();

      return true;
    } catch (error) {
      console.error('Update task status error:', error);
      return false;
    }
  }
}

module.exports = WorkflowService;